### DISCUSSION TITLE
Make graphcache a product in itself? Why not?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#1141] [Ideas]
Dear guys, you know how much I care about URQL.

Today I am writing to you because in an upcoming project that is about to start I MUST use JSONAPI and not GraphQL as backend.

What I don't want to give up is the joy of using great products like URQL.

I was wondering: would it take so long to "extrapolate" the normalized graphcache cache by decoupling it from GraphQL?

For me it would be "enough" to have a normalized cache in Svelte ([which is currently missing](https://stackoverflow.com/questions/64760734/is-there-a-normalized-data-store-for-svelte-3)) and the "stale-while-revalidate" mechanism with a minimum offline as it is today in Urql.

Do you think it is too difficult?

**Comment by JoviDeCroock** [comment]
GraphCache relies on schema-semantics to normalize the cache, this means that to normalize a response we'll use `__typename`, `id` and traverse the response given the `GraphQL AST`. This part is already very inherent to GraphQL, then we get to how `optimistic`, `updates`, `resolvers`, ... work.

I don't think it would be an easy job to port this, I think it'd be better to port something like [`react-query`](https://github.com/tannerlinsley/react-query) or [`swr`](https://github.com/vercel/swr) to Svelte.

For what would you want a `normalized cache`?

**Reply by frederikhors** [reply]
> For what would you want a normalized cache?

For the same reason I need it with urql.

**Reply by JoviDeCroock** [reply]
Yes, how would you be using that. Mainly what I mean is that a lot of the concepts would have to change drastically since there is no schema to follow for resolvers, updaters, ... so you would need a way to follow that.
This means that the only thing that might end up being reusable would be the [data layer](https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/store/data.ts) which leaves you with traversal, keying, ... left

**Reply by frederikhors** [reply]
https://jsonapi.org/ has schema.

**Reply by JoviDeCroock** [reply]
Yes, so that leaves you with all the traversal logic, .... I don't think this is an endeavour we'd like to start, we're focussing on the GraphQL space. Feel free to mix & match logic yourself in a repository if you like.

**Reply by kitten** [reply]
If anything, I’d recommend trying to convert the JSONAPI to GraphQL on the client-side or using a BFF (backend for frontend) API.

This isn’t really that novel. We even have an execute exchange for the purpose of integrating an executable schema on the frontend. And then there’s also this project: https://www.npmjs.com/package/graphql-anywhere

But ultimately, for us, this is out of scope.
